### ISSUE TITLE
External sites opened from Diaspora can access window.opener

### ISSUE THREAD

**Issue by gandaro** [diaspora/diaspora#6871]
### How to reproduce the bug?
In the Diaspora stream or in a private message on Diaspora, click on a link to a website that contains JavaScript of the following kind (e.g. [this one](https://mathiasbynens.github.io/rel-noopener/malicious.html)):
```js
if (window.opener) {
    opener.location = 'https://example.com/';
}
```
### What’s happening?
The website containing the JavaScript will open in a new tab (because of the attribute `target="_blank"`, which is set on all external links in Diaspora).

The JavaScript will be executed, and the tab containing the Diaspora stream will relocate to the URL given in the JavaScript (e.g. https://example.com/).

You can see the behavior in this animation:
![Illustration](https://cloud.githubusercontent.com/assets/811907/16174569/4113e3c4-35c7-11e6-8efd-18bab59d423d.gif)
### Why is that a problem?
No one expects their tabs to switch to completely different locations because they open a link in a new tab. This might be abused, for example, by posting links to malicious sites that send the Diaspora tab to a site mimicking the user’s Diaspora pod, showing a sign in page with a “your session timed out” message, potentially phishing credentials.
### How to prevent this?
Set `rel="noopener noreferrer"` on all (or at least user-provided) external links. See https://mathiasbynens.github.io/rel-noopener/#recommendations

**Comment by denschub** [comment]
I'm a bit concerned about the way this issue is "promoted" here and in the demo page you've linked, to be honest. I do not like people who try to make a big deal about issues that are a bit more complex. For everyone really interested in this, I suggest to read the discussion in Chrome bug tracker and all related issues.

The actual issue is not a "missing" rel attribute, but rather an implementation offset included in all browsers. The [spec is already changed](https://html.spec.whatwg.org/multipage/browsers.html#allowed-to-navigate) and the browser manufactors are working on implementing the changes:

> A browsing context A is allowed to navigate a second browsing context B if the following algorithm terminates positively:
>
> 1. If A is not the same browsing context as B, and A is not one of the ancestor browsing contexts of B, and B is not a top-level browsing context, and A's active document's active sandboxing flag set has its sandboxed navigation browsing context flag set, then abort these steps negatively.
> 2. Otherwise, if B is a top-level browsing context, and is one of the ancestor browsing contexts of A, and A's active document's active sandboxing flag set has its sandboxed top-level navigation browsing context flag set, then abort these steps negatively.
> 3. Otherwise, if B is a top-level browsing context, and is neither A nor one of the ancestor browsing contexts of A, and A's Document's active sandboxing flag set has its sandboxed navigation browsing context flag set, and A is not the one permitted sandboxed navigator of B, then abort these steps negatively.
> 4. Otherwise, terminate positively!

Basically, that resolves the issue in a sane way since CORS issues will no longer be issues.

Regarding the implementation of `rel="noopener"`: it is only [available in Chrome](https://bugs.chromium.org/p/chromium/issues/detail?id=551884). Work in Firefox [is stalled](https://bugzilla.mozilla.org/show_bug.cgi?id=1222516), [Safari](https://bugs.webkit.org/show_bug.cgi?id=155166) and [Internet Explorer](https://wpdev.uservoice.com/forums/257854-microsoft-edge-developer/suggestions/12942405-implement-rel-noopener) folks have no plans of implementing it. Actually, I've heard people discussing (you know, I'm active in such channels) about the usefulness of `rel="noopener"` after CORS rules are implemented, so it may become deprecated again. Or it [might change to something else](https://wiki.whatwg.org/wiki/Links_to_Unrelated_Browsing_Contexts).

Anyway, since I do not really want to have extensive discussions about this and the actual fix is rather easy, I've opened #6872. In any case, some rel attributes will not hurt, even if they are deprecated, so who cares.

**Comment by gandaro** [comment]
According to Chrome’s Charlie Reis the HTML specification [used to prohibit a cross-origin page from navigating their openers](https://bugs.chromium.org/p/chromium/issues/detail?id=168988#c3), and the current wording [“now allows cross-origin openers to be navigated”](https://bugs.chromium.org/p/chromium/issues/detail?id=168988#c12). That sounds contradictory to your statement:

>Basically, that resolves the issue in a sane way since CORS issues will no longer be issues.

Anyway, thank you for fixing the issue.
